### PR TITLE
[Copy update to assert for overlap storage at destination buffer]

### PR DESCRIPTION
**Pull request by jjsjann123**
Assert that there's no multiple written-to to a single memory location, which
caused corrupted output.
Fixed batched matrix trlu logic, which relies on the previous copy behavior to
support tensors with stride 0 at leading dimension.

This fixes the issue proposed at: https://github.com/pytorch/pytorch/issues/23063

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -4579,6 +4581,9 @@ def test_copy_broadcast(self):
         torch.zeros(5, 6).copy_(torch.zeros(6))
         self.assertRaises(RuntimeError, lambda: torch.zeros(5, 6).copy_(torch.zeros(30)))
 
+    def test_copy_many_to_one(self):
+        self.assertRaises(RuntimeError, lambda: torch.zeros(1, 6).expand(5, 6).copy_(torch.zeros(5, 6)))
```
Comment (VitalyFedyunin) [inline test/test_torch.py]: super nit: such one lines makes it hard to understand what is supposed to raise an exception here

File: aten/src/ATen/native/LinearAlgebraUtils.h [hunk 2]
```diff
@@ -44,11 +44,25 @@ static inline int64_t matrixStride(const Tensor& batched_matrices) {
...
+
 /* Checks a necessary property for the triu and tril implementations, hence the name.
  * Here batch contiguity is checked for tensors with greater than 4 dimensions.
  * Contiguous tensors and tensors with less than 3 dimensions pass this check
  */ 
-static inline std::tuple<bool, Tensor> checkTrilTriuBatchContiguous(const Tensor& tensor) {
+static inline std::tuple<bool, Tensor> checkTrilTriuBatchContiguous(const Tensor& tensor, bool allow_zero_stride) {
   // Complete contiguity is the most desired property, which is why
   // we return true if the tensor is contiguous
   if (tensor.is_contiguous()) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/LinearAlgebraUtils.h]: Do we ever have situation when is_contiguous tensor doesn't satisfy `tensor.strides() == default_strides_for_size`. As I can see `defaultStrides` and `conmute_contiguous` using exactly same strides formula.

**Comment by VitalyFedyunin** [comment]
cc @pbelevich

File: aten/src/ATen/native/LinearAlgebraUtils.h (hunk 2 above)
Comment (jjsjann123) [reply aten/src/ATen/native/LinearAlgebraUtils.h]: Here we go. :)
```
In [7]: x = torch.randn(2, 10, 2)                                                                                                    

In [8]: y = x[::2, :, :]                                                                                                             

In [9]: y.is_contiguous()                                                                                                            
Out[9]: True

In [10]: y.size()                                                                                                                    
Out[10]: torch.Size([1, 10, 2])

In [11]: y.stride()                                                                                                                  
Out[11]: (40, 2, 1)```

File: test/test_torch.py (hunk 1 above)
Comment (jjsjann123) [reply test/test_torch.py]: Agreed.
Will comment for the test case.

File: aten/src/ATen/native/LinearAlgebraUtils.h (hunk 2 above)
Comment (jjsjann123) [reply aten/src/ATen/native/LinearAlgebraUtils.h]: dimension with size 1 always got me as well

File: aten/src/ATen/native/Copy.cpp [hunk 3]
```diff
@@ -131,6 +132,7 @@ Tensor & copy_(Tensor & self, const Tensor & src, bool non_blocking) {
   }
 
   auto iter = TensorIterator();
+  assert_no_internal_overlap(self);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Copy.cpp]: You can do `iter.check_and_add_output(self)` instead

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by vincentqb** [comment]
@VitalyFedyunin -- I'll let you land this PR from the "needs land" queue.

**Comment by ezyang** [comment]
Waiting on CI after merge.

**Pull request closed without merging**